services/command-router-base/src/main/java/org/eclipse/hono/commandrouter/impl/CommandRouterServiceImpl.java

```diff
@@ -209,53 +220,94 @@ private Duration getSanitizedLifespan(final Duration lifespan) {
-                        .addReference(References.CHILD_OF, parentSpan.context())
-                        .withTag(TracingHelper.TAG_TENANT_ID, tenantId)
-                        .start();
-                tenantClient.get(tenantId, span.context())
-                    .map(tenantObject -> commandConsumerFactories.getClient(tenantObject))
-                    .map(factory -> factory.createCommandConsumer(tenantId, span.context()))
-                    .onSuccess(ok -> {
-                        span.log("successfully created command consumer");
-                        processedTenants.add(tenantId);
-                    })
-                    .onFailure(t -> {
-                        TracingHelper.logError(span, "failed to create command consumer", t);
-                        if (t instanceof ServerErrorException) {
-                            // add to end of queue in order to retry at a later time
-                            span.log("marking tenant for later re-try to create command consumer");
-                            tenantsToEnable.add(tenantId);
-                        }
-                    })
-                    .onComplete(r -> {
-                        span.finish();
-                    });
-            }
-            processTenantQueue(processedTenants, parentSpan);
-        });
+
+        final Span span = tracer.buildSpan("re-enable command routing for tenant")
+                .addReference(References.FOLLOWS_FROM, tracingContext)
+                .withTag(TracingHelper.TAG_TENANT_ID, attempt.one())
+                .start();
+        final var logEntries = new HashMap<String, Object>(2);
+        logEntries.put("attempt#", attempt.two());
+        tenantClient.get(attempt.one(), span.context())
+            .map(tenantObject -> commandConsumerFactories.getClient(tenantObject))
+            .map(factory -> factory.createCommandConsumer(attempt.one(), span.context()))
+            .onSuccess(ok -> {
+                logEntries.put(Fields.MESSAGE, "successfully created command consumer");
+                span.log(logEntries);
+                processedTenants.add(attempt.one());
+            })
+            .onFailure(t -> {
+                logEntries.put(Fields.MESSAGE, "failed to create command consumer");
+                logEntries.put(Fields.ERROR_OBJECT, t);
+                TracingHelper.logError(span, logEntries);
+                if (t instanceof ServerErrorException) {
+                    // add to end of queue in order to retry at a later time
+                    LOG.info("failed to create command consumer [attempt#: {}]", attempt.two(), t);
+                    span.log("marking tenant for later re-try to create command consumer");
+                    // with the maximum delay of 10s before re-trying,
+                    // the attempt number will increase to 
+                    tenantsToEnable.addLast(Pair.of(attempt.one(), attempt.two() + 1));
```

Comment:
I would suggest checking first whether `tenantsToEnable` already contains an entry for that tenant (the entry could have been added while processing above took place or during the timer period) and removing that entry first.

Reply:
I have pushed some changes